### PR TITLE
travis: add boringssl build

### PR DESCRIPTION
**Pull request by bagder**

### CODE REVIEW

**Comment by bagder** [comment]
@davidben! Do you have a suggested tag/tag system we should use when we run test builds of curl/boringssl, or is using borginssl's master branch reasonably safe?

File: .travis.yml [hunk 1]
```diff
@@ -108,6 +112,20 @@ before_script:
                   sudo make install
               )
           fi
+          if [ "$TRAVIS_OS_NAME" = linux -a "$BORINGSSL" ]; then
+            (git clone https://boringssl.googlesource.com/boringssl &&
+            cd boringssl &&
```
Comment (wmark) [inline .travis.yml]: Some revisions need this:

```bash
sed -i -e '/^if (BUILD_SHARED_LIBS)/i\add_definitions(-DBORINGSSL_ENABLE_DHE_TLS)' CMakeLists.txt &&
```

File: .travis.yml [hunk 2]
```diff
@@ -108,6 +112,20 @@ before_script:
...
+            mkdir build &&
+            cd build &&
+            cmake -DCMAKE_INSTALL_PREFIX:PATH=$HOME/build-boringssl .. &&
+            make &&
```
Comment (wmark) [inline .travis.yml]: Using *cmake* with this option is probably what most maintainers use:
```bash
cmake -DCMAKE_BUILD_TYPE=release …
```

… and then you could decrease compilation time by (replace *ninja* by *make*; don't forget *libdecrepit.a*):

```bash
ninja crypto/libcrypto.a ssl/libssl.a decrepit/libdecrepit.a
```

FYI, you could build everything as shared libraries, but I don't think this makes a difference for the tests.

File: .travis.yml [hunk 3]
```diff
@@ -108,6 +112,20 @@ before_script:
                   sudo make install
               )
           fi
+          if [ "$TRAVIS_OS_NAME" = linux -a "$BORINGSSL" ]; then
+            (git clone https://boringssl.googlesource.com/boringssl &&
```
Comment (wmark) [inline .travis.yml]:
```bash
git clone --depth=1 …
```

… if you don't rewind to the latest tag or any blessed revision, using a shallow clone will be faster.

File: .travis.yml (hunk 2 above)
Comment (bagder) [reply .travis.yml]: > libdecrepit.a

What's that and what do we need that for?

File: .travis.yml (hunk 3 above)
Comment (bagder) [reply .travis.yml]: good idea!

File: .travis.yml (hunk 2 above)
Comment (wmark) [reply .travis.yml]: »Old« algorithms are collected in that library. For example, DSA and RC4.

Comment (bagder) [reply .travis.yml]: And regarding ninja: I dread to add more dependencies (problems) to the annoying travis environment. Besides, make vs ninja can't make *that* big difference when building from scratch in a single-core one-by-one build surely?

Comment (wmark) [reply .travis.yml]: Never used *make* here, but I think *make* accepts what-to-build as arguments like *ninja* does.

(I guess you want to use `make -j$(nproc) …` here.)

Comment (wmark) [reply .travis.yml]: Tried it, `make -j$(nproc)` works, using *make* you cannot compile a subset of files like when using *ninja*.

Comment (bagder) [reply .travis.yml]: > using make you cannot compile a subset of files like when using ninja.

so you're saying the generated makefile builds too many files?

Comment (wmark) [reply .travis.yml]: Yes, but there's no harm in that. Other than some cpu cycles.

Comment (bagder) [reply .travis.yml]: we can consider changing that in phase two, just now I want to focus on getting this to work...

**Comment by bagder** [comment]
Hm, the build works now, but we get memory leakage from inside boringssl. I'll try to reproduce this locally and see what's going on.

**Comment by bagder** [comment]
It seems to be a false positive on thread-local memory, possibly caused by a buggy/old valgrind version. Since we already run many tests with valgrind enabled, I figured switching it off for the boringssl build is an okay work-around.

**Comment by wmark** [comment]
Remains the issue of what revision to pick for tests.

In the ticket leading up to this one I suggested to not go with HEAD because sometimes breaking changes occur (which sometimes are reverted, sometimes not).

**Comment by bagder** [comment]
I asked David Benjamin and he said:

> If you care about a "reasonably up-to-date" BoringSSL (which is also how we typically do things) just tracking master probably makes sense then.

... so I figure we can try doing that and see how it goes.

**Comment by wmark** [comment]
Good enough for me. Thanks!

**Comment by bagder** [comment]
The build looks red here only because I cancelled all the non-boringssl builds. I'll proceed and merge this into master now...

**Pull request closed without merging**